File: trend_detection/processors/trend_detection_processor.py
from datetime import datetime, timedelta
import json
import time
import pytz

from pyflink.common import Types, Row
from pyflink.datastream.functions import KeyedProcessFunction, RuntimeContext
from pyflink.datastream.state import ValueStateDescriptor

from trend_detection_embeddings import TrendDetectorEmbeddings
import utils


TREND_CREATED = 'TREND_CREATED'
TREND_STATS = 'TREND_STATS'


class TrendDetectionProcessor(KeyedProcessFunction):
    def __init__(self, window_minutes=5):
        self.td = None
        self.scheduled_windows = None
        self.window_minutes = window_minutes

    def open(self, runtime_context: RuntimeContext):
        self.td = TrendDetectorEmbeddings(window_minutes=self.window_minutes)

        self.scheduled_windows = runtime_context.get_state(
            ValueStateDescriptor("scheduled_windows", Types.LIST(Types.LONG()))
        )

    def _schedule_window_end_callback(self, ctx, timestamp):
        msg_timestamp = datetime.fromisoformat(timestamp)
        window_start = utils.timestamp_to_window_start(msg_timestamp, self.window_minutes)
        window_end = int((window_start + timedelta(minutes=self.window_minutes)).timestamp())

        scheduled = self.scheduled_windows.value() or []
            
        if window_end not in scheduled:
            ctx.timer_service().register_event_time_timer(window_end)
            scheduled.append(window_end)
            self.scheduled_windows.update(scheduled)

    def process_element(self, value, ctx: 'KeyedProcessFunction.Context'):
        self._schedule_window_end_callback(ctx, value.timestamp)

        detected_trends = self.td.process_message(
            value.text, 
            value.timestamp,
            value.lat,
            value.lon,
            time.time(), 
            debug_trend_id=value.d_trend_id, 
            debug_location_id=value.d_location_id)

        for trend in detected_trends:
            print(f"detected: {trend.id}")
            yield Row(
                trend_event=TREND_CREATED,
                trend_id=trend.id,
                keywords=', '.join(trend.keywords),
                location_id=ctx.get_current_key(),
                info=f"location:{trend.debug_location_ids}; trends:{trend.debug_trend_ids}",
            )

    def on_timer(self, timestamp: int, ctx: 'KeyedProcessFunction.OnTimerContext'):
        scheduled = self.scheduled_windows.value()
        if scheduled is not None:
            scheduled.remove(timestamp)
            self.scheduled_windows.update(scheduled)

        for el in []:
            yield el
            
        location_id = ctx.get_current_key()
        window_end = datetime.fromtimestamp(timestamp).replace(tzinfo=pytz.UTC)
        window_start = window_end - timedelta(minutes=self.window_minutes)

        # print(location_id, window_start, self.td.trends.keys())

        for trend in self.td.trends.values():
            stats = trend.stats.get_timestamp_stats(window_start)
            if not stats:
                continue
            
            # print(location_id, window_start, window_start.isoformat() in list(trend.stats.stats.keys()), json.dumps(stats))

            yield Row(
                event_type=TREND_STATS,
                location_id=location_id,
                trend_id=trend.id,
                window_start=window_start.isoformat(),
                window_end=window_end.isoformat(),
                stats=json.dumps(trend.get_timestamp_stats(window_start)),
            )

File: trend_detection/java/src/main/java/com/trendpulse/processors/TrendDetectionProcessor.java
package com.trendpulse.processors;

import org.apache.commons.lang3.ArrayUtils;
import org.apache.flink.api.common.state.ListState;
import org.apache.flink.api.common.state.ListStateDescriptor;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;

import com.azure.ai.openai.OpenAIClient;
import com.azure.ai.openai.OpenAIClientBuilder;
import com.azure.core.credential.AzureKeyCredential;
import com.trendpulse.TrendDetector;
import com.trendpulse.items.InputMessage;
import com.trendpulse.items.TrendDetected;
import com.trendpulse.lib.TimeUtils;
import com.trendpulse.lib.TrendNameGenerator;
import com.trendpulse.schema.*;


public class TrendDetectionProcessor extends KeyedProcessFunction<Tuple2<Integer, String>, InputMessage, TrendEvent> {

    private static final long serialVersionUID = 1L;

    private final String socketPath;
    private final int trendStatsWindowMinutes;
    private transient Map<Tuple2<Integer, String>, TrendDetector> trendDetectorsMap;
    private transient ListState<Long> scheduledWindows;

    private transient TrendNameGenerator trendNameGenerator;

    public TrendDetectionProcessor(String socketPath, int trendStatsWindowMinutes) {
        this.socketPath = socketPath;
        this.trendStatsWindowMinutes = trendStatsWindowMinutes;
    }

    @Override
    public void open(Configuration conf) throws Exception {
        this.trendDetectorsMap = new HashMap<>();
        this.scheduledWindows = getRuntimeContext().getListState(
            new ListStateDescriptor<>("scheduled_windows", Long.class)
        );

        trendNameGenerator = new TrendNameGenerator();
    }

    private void scheduleWindowEndCallback(Context ctx, OffsetDateTime datetime) throws Exception {
        Instant windowStart = TimeUtils.timestampToWindowStart(datetime.toInstant(), trendStatsWindowMinutes);
        Instant windowEnd = windowStart.plus(trendStatsWindowMinutes, ChronoUnit.MINUTES);
        long windowEndMillis = windowEnd.toEpochMilli();

        // Check if window end is already scheduled
        boolean isScheduled = false;
        for (Long scheduled : scheduledWindows.get()) {
            if (scheduled == windowEndMillis) {
                isScheduled = true;
                break;
            }
        }

        if (!isScheduled) {
            ctx.timerService().registerEventTimeTimer(windowEndMillis);
            scheduledWindows.add(windowEndMillis);
        }
    }

    @Override
    public void processElement(InputMessage message, Context ctx, Collector<TrendEvent> out) 
            throws Exception {
                
        Integer locationId = ctx.getCurrentKey().f0;
        String topic = ctx.getCurrentKey().f1;
        scheduleWindowEndCallback(ctx, message.getDatetime());

        // if (locationId != 16) {
        //     return;
        // }

        if (message.getDTrendId() != 1) {
            return;
        }

        if (!trendDetectorsMap.containsKey(ctx.getCurrentKey())) {
            trendDetectorsMap.put(ctx.getCurrentKey(), new TrendDetector(locationId, topic, socketPath, trendStatsWindowMinutes));
        }

        TrendDetector trendDetector = this.trendDetectorsMap.get(ctx.getCurrentKey());
        TrendDetector.ProcessingResult result = trendDetector.processMessage(message, ctx.timestamp());
        
        String events = "";

        if (result != null ) {
            for (TrendDetected trend : result.getActivatedTrends()) {
                events += " TREND_ACTIVATED ";

                List<CharSequence> keywords = new ArrayList<>(trend.getKeywords());
                List<CharSequence> sampleMessages = trend.getMessages().stream().limit(10).map(m -> m.getText()).collect(Collectors.toList());

                TrendEvent event = new TrendEvent(
                    EventType.TREND_ACTIVATED,
                    trend.getId(),
                    TrendType.TREND_TYPE_LOCAL,
                    trend.getTopic(),
                    new LocalTrendInfo(trend.getLocationId()),
                    new TrendActivatedInfo(
                        trendNameGenerator.generateTrendName(keywords, sampleMessages),
                        keywords,
                        Arrays.asList(ArrayUtils.toObject(trend.getCentroid())),
                        sampleMessages
                    ));
                
                out.collect(event);
            }
        }

        for (TrendDetected trend : result.getDeActivatedTrends()) {
            
            events += " TREND_DEACTIVATED ";
            TrendEvent event = new TrendEvent(
                EventType.TREND_DEACTIVATED,
                trend.getId(),
                TrendType.TREND_TYPE_LOCAL,
                trend.getTopic(),
                new LocalTrendInfo(trend.getLocationId()),
                null
            );
            
            out.collect(event);
        }

        // Files.write(
        //     // Paths.get("local_messages_5_min.csv"),
        //     Paths.get("kafka_messages_4_partitions.csv"),
        //     String.format(
        //         "%d, %d, %d, %s\n", 
        //         ctx.timestamp(), 
        //         ctx.timerService().currentWatermark(), 
        //         (ctx.timestamp() - ctx.timerService().currentWatermark()) / 1000 / 60,
        //         events
        //     ).getBytes(),
        //     StandardOpenOption.CREATE,
        //     StandardOpenOption.APPEND
        // );
    }

    @Override
    public void onTimer(long timestamp, OnTimerContext ctx, Collector<TrendEvent> out) throws Exception {
        // Remove this timestamp from scheduled windows
        List<Long> currentScheduled = new ArrayList<>();
        scheduledWindows.get().forEach(currentScheduled::add);
        currentScheduled.remove(timestamp);
        scheduledWindows.clear();
        
        currentScheduled.forEach(t -> {
            try {
                scheduledWindows.add(t);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });

        Instant windowEnd = Instant.ofEpochMilli(timestamp);
        Instant windowStart = windowEnd.minus(trendStatsWindowMinutes, ChronoUnit.MINUTES);

        TrendDetector trendDetector = trendDetectorsMap.get(ctx.getCurrentKey());
        
        if (trendDetector != null) {
            for (TrendDetected trend : trendDetector.getTrends()) {

                // Do not emit stats too early
                if (windowEnd.toEpochMilli() < trend.getCreatedAt()) {
                    continue;
                }

                TrendEvent event = new TrendEvent(
                    EventType.TREND_STATS,
                    trend.getId(),
                    TrendType.TREND_TYPE_LOCAL,
                    trend.getTopic(),
                    new LocalTrendInfo(trend.getLocationId()),
                    new TrendStatsInfo(trend.getStats().getWindowStats(windowStart))
                );

                out.collect(event);
            }
        }
    }
}